### DISCUSSION TITLE
Library mode vs Framework mode

### DISCUSSION THREAD

**Discussion by AsuraKev** [remix-run/react-router#12423] [Q&A]
Hi there, I was going through the v7 docs and notice that v7 marks the boundary between library mode and framework mode.
Most of us most likely would have existing projects scaffolded from vite command, does that mean we need to create a new project via react-router cli in order to opt into framework mode? 

I notice framework mode is packed with a lot more features than library mode, however the v6 way of declaring routes has also changed? e.g

```ts
const router = createBrowserRouter([
  {
    path: "/",
    element: <Root />,
    loader: rootLoader,
    children: [
      {
        path: "team",
        element: <Team />,
        loader: teamLoader,
      },
    ],
  },
]);
```

and v7 docs suggest to declare routes via Route Components?

```
<BrowserRouter>
    <Routes>
      <Route path="/" element={<App />} />
    </Routes>
  </BrowserRouter>
```

Can someone please explain to me whats going on and which mode we should be going for if existing projects? 

thanks

**Comment by discoverlance-com** [comment]
I think you can use any routing mode that suits your project.

## Library Mode

This is for when you want to use react router in another project that already has a react setup, this can be a vite project for example. So use library mode if you have or want to use vite.

If you are not interested in using vite and you want to just use react router, then you can use the framework mode.

## Framework Mode

The framework mode should support all the routing features you will get from the library mode. Only that you declare routes a bit more differently. There are over 3 ways I know of and you can see information about them below.


### Route Config

You can use the route config to configure your routes. You can find the information here, [Routing](https://reactrouter.com/start/framework/routing). It discusses how to use the route config to configure your routes. It supports layouts, nested routes, route prefixes, loaders,actions and other features you can check out in the document.

### File based Routing

File based routing allows you to configure your routes using files as your routes like you have in remix or nextjs.

Follow this documentation for how to set this up with react router (https://reactrouter.com/how-to/file-route-conventions). You will have to update your routes config file with the fs routes package and then follow the document on how to write your file based routes. This also supports layouts, nested routes, route prefixes , loaders, actions and other router features.

### Component Routes

Now, [component routes](https://reactrouter.com/how-to/file-route-conventions) in the react router framework is not exactly a routing strategy for your whole application. It just allows you to render a page and then have a section of it render different components based on the url. It does not support the full react router route module features like data loading, actions... It's to be used when you just want to render a different react component on the page based on the url. But they can of course use any feature in react router that a normal component can use like `useNavigation` or `useLocation`... They are meant to be added to a route (either file based or using the route config).


Whether you are using the library mode or framework mode with file based routing or route configuration, they all allow you to use the react router [route modules](https://reactrouter.com/start/framework/route-module), so you have access to all features such as actions, loaders, ErrorBoundary, links, meta...

**Reply by sergiodxa** [reply]
> If you are not interested in using vite and you want to just use react router, then you can use the framework mode

React Router framework mode depends on a Vite plugin, so if you're not interested in using Vite you can't use framework mode, you have to use library mode

React Router library mode doesn't depend on any compiler so it can be used with Vite, Parcel, webpack, etc.

**Reply by GabenGar** [reply]
Judging from the amount of disinfo (while also completely missing the mark of discussion about SPAs specifically) and excessive formatting, I assume it's a chatgpt response.

**Reply by turnerguo** [reply]
> > If you are not interested in using vite and you want to just use react router, then you can use the framework mode
> 
> React Router framework mode depends on a Vite plugin, so if you're not interested in using Vite you can't use framework mode, you have to use library mode
> 
> React Router library mode doesn't depend on any compiler so it can be used with Vite, Parcel, webpack, etc.

why should react-router depends on one bundler like Vite? what about webpack/rspack some else? is there any plan to support webpack/rspack(rsbuild) later?

**Reply by sergiodxa** [reply]
There's no public plan so far to publish official plugins for other bundlers, frameworks typically choose a bundler, most now use Vite, Next.js used webpack for a long time end now uses a custom one, so it's not strange for RR framework mode to choose one bundler, and Vite is really popular.

**Comment by ArthurGuez** [comment]
Hello there,

I'm also confused.
I'm using RR6.28.0 with createBrowserRouter and route loaders.
Does that mean that if I update to RR7, I have to use framework mode and the Vite plugin to make everything work as before?
Thanks

**Reply by sergiodxa** [reply]
You can upgrade to RRv7 and still use createBrowserRouter and route loaders, the function still exists https://api.reactrouter.com/v7/functions/react_router.createBrowserRouter.html

**Reply by GabenGar** [reply]
Are there plans to expand descriptions to the level of v6 docs? Just plain type info is pretty useless outside of IDE.

**Reply by sergiodxa** [reply]
I'm not part of the team, but I know Brooks is working on a better tutorial and more docs

**Comment by EPurwanto** [comment]
Why does the Framework Installation page not include instructions for enabling it in an existing project? It just says "You can [view the template on GitHub](https://github.com/remix-run/react-router-templates/tree/main/default) to see how to manually set up your project." That's not particularly helpful.

There's even already instructions for how to set it up in the "Framework Adoption from RouterProvider/Component Routes" sections, they just assume you've already got a Library installation set up.

**Comment by aramis-it** [comment]
Yes, Me too wondering. Should I use library or framework in the future?

**Reply by sergiodxa** [reply]
If possible using framework mode would be better as you won't have to do many things by yourself.

Library mode is more for people who already uses RR and maybe has a custom build process and can't switch to framework mode, a new app could go directly to framework mode unless there's any other requirement that prevents that (like maybe an architecture decision of the app itself)

**Reply by deronendless** [reply]
Are there examples of framework mode?

**Reply by sergiodxa** [reply]
The templates [here](https://github.com/remix-run/react-router-templates) are using framework mode.

**Comment by kno-raziel** [comment]
note: not sure if this is the right place...

Will the framework mode be supported by other bundlers than Vite ?, I noticed that Tanstack Router supports different bundlers while react router only Vite 🫠

![image](https://github.com/user-attachments/assets/f8cdf9b1-f46d-403d-923f-9db52bbff3d5)


I work in a project were we are using RRDv6, we would like to upgrade to V7 and use framework features but we use rspack....

**Reply by GabenGar** [reply]
Well it's time to migrate to vite then or stick to data/declarative mode.

**Reply by sergiodxa** [reply]
There's a Rspack port of the plugin maintained by the authors of Repack https://github.com/rspack-contrib/rsbuild-plugin-react-router, I never tried it but in theory it supports the same features

But there's no public plan to support other build tools

**Comment by turnerguo** [comment]
already dropped v7, since it makes a wrong decision!🙄


Thanks,
Turner
 
---- Replied Message ----
| From | ***@***.***> |
| Date | 04/24/2025 01:11 |
| To | remix-run/react-router ***@***.***> |
| Cc | Turner ***@***.***>,
Comment ***@***.***> |
| Subject | Re: [remix-run/react-router] Library mode vs Framework mode (Discussion #12423) |

note: not sure if this is the right place...

Will the framework mode be supported by other bundlers than Vite ?, I noticed that Tanstack Router supports different bundlers while react router only Vite

image.png (view on web)

I work in a project were we are using RRDv6, we would like to upgrade to V7 and use framework features but we use rspack....

**Reply by jacob-ebey** [reply]
React Router v7 did not change how you use it if you were using v6. All it did was add a (optional) Vite plugin that enforces some conventions. YOU CAN CONTINUE TO USE IT EXACTLY THE SAME AS YOU HAVE BEEN.

**Reply by jacob-ebey** [reply]
We even publish `react-router-dom` so you don't have to change your application code if you don't want to: https://www.npmjs.com/package/react-router-dom

HOW DO WE STOP THIS NARRATIVE / how did you end on this idea?

**Reply by GabenGar** [reply]
Upgrade path from V6 literally says to delete that package:
https://reactrouter.com/upgrading/v6#upgrade-to-v7
The issue was created when the docs had only Library/Framework separation instead of Framework/Data/Declarative.
Plus the docs were like "here is extremely barebones setup for framework mode (aka runtime server render (aka something only a limited subset of react router users need)) and a very basic example of library mode (aka classic client-side render (aka majority of react router users need)) which doesn't even mention data routers at all". Safe to say it was confusing both for people migrating from V6 and arriving fresh to V7. I guess it wasn't confusing for remix people, but only because they are accustomed to dealing with hybrid render stack problems.

**Comment by MichaelHao25** [comment]
Framework mode  use redux issue？

**Comment by hb20007** [comment]
I'm confused about React Router being called a "full-stack framework" as of v7. So, can you write APIs with it? What if I don't want this feature? Is the "library mode" more lightweight?

**Reply by sergiodxa** [reply]
If you're using React Router Framework mode you could totally write an API with it, an API route is any normal React Router route file without a component being exported, only with a loader (GET handler) and action (non-GET handler), this is called a [Resource Route](https://reactrouter.com/how-to/resource-routes) as it's not limited to JSON but any response.

```ts
import type { Route } from "./+types/resource";

export function loader(_: Route.LoaderArgs) {
  return Response.json({ message: "I handle GET" });
}

export function action(_: Route.ActionArgs) {
  return Response.json({
    message: "I handle everything else",
  });
}
```

And you can use `await request.json()` to receive the request body on an action for an API.

---

If you don't want to use framework mode you can use data mode, there's an guide on [how to pick a mode](https://reactrouter.com/start/modes) and the difference between them.
